### PR TITLE
Better IAST::format() interface

### PR DESCRIPTION
**Pull request by vitlibar**
### Changelog category (leave one):
- Not for changelog (changelog entry is not required)

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Better IAST::format() interface:
- `IAST::formatImpl()` is now protected, and `IAST::format()` is called instead.
- class `IAST::FormattingBuffer` was added as preparation for https://github.com/ClickHouse/ClickHouse/pull/45649

This is part 2 of https://github.com/ClickHouse/ClickHouse/pull/45649

### CODE REVIEW

**Comment by vitlibar** [comment]
CI failures are unrelated

File: src/Parsers/ASTQueryWithOutput.h [hunk 1]
```diff
@@ -18,13 +18,11 @@ class ASTQueryWithOutput : public IAST
     bool is_into_outfile_with_stdout = false;
     bool is_outfile_append = false;
     bool is_outfile_truncate = false;
-    ASTPtr format;
+    ASTPtr format_ast;
```
Comment (vitlibar) [inline src/Parsers/ASTQueryWithOutput.h:21]: I renamed `ASTQueryWithOutput::format` to `format_ast` to avoid a confusion with `IAST::format()`

**Comment by divanik** [comment]
Why do we need this change?

**Comment by vitlibar** [comment]
> Could you explain, please, what for we need this change?

There is another PR https://github.com/ClickHouse/ClickHouse/pull/45649. In that PR class [IAST::FormattingBuffer](https://github.com/ClickHouse/ClickHouse/pull/45649/files#diff-b8e3361e1eecb9d87beb580ab7bc3ce662aada156e5f7fe8a1add5faaf253c8dR225) was added and all the `format()` functions in `IAST`-derived class were rewritten to use that class. PR https://github.com/ClickHouse/ClickHouse/pull/45649 has become very big and very conflicted. I want to apply changes from that PR step by step, that's how this PR appeared.

**Review by divanik — APPROVED** [review]

**Pull request merged**